packages/vite/src/node/plugins/html.ts

```diff
@@ -116,14 +116,261 @@ export function addToHTMLProxyTransformResult(
   htmlProxyResult.set(hash, code)
 }
 
-// this extends the config in @vue/compiler-sfc with <link href>
-export const assetAttrsConfig: Record<string, string[]> = {
-  link: ['href'],
-  video: ['src', 'poster'],
-  source: ['src', 'srcset'],
-  img: ['src', 'srcset'],
-  image: ['xlink:href', 'href'],
-  use: ['xlink:href', 'href']
+export interface HtmlAssetSource {
+  tag: string
+  attributes: string[]
+  type: 'src' | 'srcset'
+  filter?: (attribute: string, value: string, node: ElementNode) => boolean
+}
+
+type AttributeNodeContext = {
+  context: {
+    sourceType: 'src' | 'srcset'
+  }
+}
+
+export const defaultHtmlAssetSources: HtmlAssetSource[] = [
+  {
+    tag: 'audio',
+    type: 'src',
+    attributes: ['src']
+  },
+  {
+    tag: 'embed',
+    type: 'src',
+    attributes: ['src']
+  },
+  {
+    tag: 'img',
+    type: 'src',
+    attributes: ['src']
+  },
+  {
+    tag: 'img',
+    type: 'srcset',
+    attributes: ['srcset']
+  },
+  {
+    tag: 'input',
+    type: 'src',
+    attributes: ['src']
+  },
+  {
+    tag: 'object',
+    type: 'src',
+    attributes: ['data']
+  },
+  {
+    tag: 'source',
+    type: 'src',
+    attributes: ['src']
+  },
+  {
+    tag: 'source',
+    type: 'srcset',
+    attributes: ['srcset']
+  },
+  {
+    tag: 'track',
+    type: 'src',
+    attributes: ['src']
+  },
+  {
+    tag: 'video',
+    type: 'src',
+    attributes: ['poster', 'src']
+  },
+  {
+    tag: 'image',
+    type: 'src',
+    attributes: ['href', 'xlink:href']
+  },
+  {
+    tag: 'use',
+    type: 'src',
+    attributes: ['href', 'xlink:href']
+  },
+  {
+    tag: 'link',
+    type: 'src',
+    attributes: ['href'],
+    filter: (attribute, value, node) => {
+      const relProp = findProp(node, 'rel')
+      if (
+        relProp &&
+        relProp.type === NodeTypes.ATTRIBUTE &&
+        relProp.value?.content &&
+        [
+          'stylesheet',
+          'icon',
+          'shortcut icon',
+          'mask-icon',
+          'apple-touch-icon',
+          'apple-touch-icon-precomposed',
+          'apple-touch-startup-image',
+          'manifest',
+          'prefetch',
+          'preload'
+        ].includes(relProp.value.content)
+      ) {
+        return true
+      }
+
+      const itempropProp = findProp(node, 'itemprop')
+      if (
+        itempropProp &&
+        itempropProp.type === NodeTypes.ATTRIBUTE &&
+        itempropProp.value?.content &&
+        [
+          'image',
+          'logo',
+          'screenshot',
+          'thumbnailurl',
+          'contenturl',
+          'downloadurl',
+          'duringmedia',
+          'embedurl',
+          'installurl',
+          'layoutimage '
+        ].includes(itempropProp.value.content)
+      ) {
+        return true
+      }
+
+      return false
+    }
```

Comment:
I think this can also be configured here. I don't know why it should be turned into a method. I would appreciate it if you could tell me

```js
{
  rel: ['stylesheet','icon','shortcut icon','mask-icon','apple-touch-icon','apple-touch-icon-precomposed','apple-touch-startup-image','manifest','prefetch','preload'],
  itemprop: ['image','logo','screenshot','thumbnailurl','contenturl','downloadurl','duringmedia','embedurl','installurl','layoutimage']
}
```

Reply:
I didn't give much thought to how it should be configured. I basically replicated the API from [html-loader](https://webpack.js.org/loaders/html-loader/), as proposed by https://github.com/vitejs/vite/issues/7362#issuecomment-1071989522. Furthermore, the approach is most likely a matter of personal preference. If the array-based configuration fits better into vites config schema then that's how it should be done, so as not to create unnecessary complexity where it isn't required.

Reply:
Since the processing I've seen is all the same, I'm guessing it can be done this way. 🫣  This reduces a lot of code.

Reply:
I would also prefer we simplify the code. But the functional form would allow for a more flexible configuration point since the PR is exposing `build.htmlAssetSources`. I'm not sure we would like to expose this complexity it is better to leave it to userland. Maybe we could only correct the list as a first step. Added the issue for discussion in a future team meeting.

Reply:
From my experience building [svelte-preprocess-import-assets](https://github.com/bluwy/svelte-preprocess-import-assets), this is actually the only way to ensure correctness (with a filter). I tried simplifying it but it isn't quite achievable if we follow the html spec.

Reply:
> since the PR is exposing `build.htmlAssetSources`

It doesn't look like the PR expose it now (maybe I've missed a discussion) 🤔

Reply:
Ya, the change to remove it was force pushed. We can still discuss it in the meeting, the PR was exposing `build.htmlAssetSources` and letting you define whatever you want instead of the default `defaultHtmlAssetSources` array. The API was exactly the one of that array.

Reply:
Sorry, my mistake. I discovered some old incorrectly merged changes while rebasing the branch earlier. However, I have the changes that expose `build.htmlAssetSources` saved locally. I'll commit them later so that the discussion has a point of reference.